### ISSUE TITLE
Expected throughput broker service ?

### ISSUE THREAD

**Issue by johnkattenhorn** [loekd/ServiceFabric.PubSubActors#19]
We are currently experimenting with PubSub and performing some load tests to assess it's suitability for a project.

Although we use Actors for managing our State we opted to implement the Broker using the brokerservice. With one or two users we get approximately 300ms response times from publishing an event.

When we up the load to around 50 users all publishing as quickly as they can through our API the performance of the broker dropoff quite a bit. The same test as before get get approximately 4.9 seconds response.

We are using 5 node cluster with 10 partitions for the BrokerService, I see mention of using BrokerActors in `large scale` system but I couldn't find a definition of large.

Are we seeing the expected behaviour of the broker service and should we switch to Actors to gain better performance ?

**Comment by loekd** [comment]
Hi John, yes the documentation could use a little work. I recommend using the BrokerService like you currently are. 

- How many message types do you use? (partitioning works based on message type hashes)
- Are you using the latest version of the package?
- How large is the message that you broadcast?
- Do you see anything in diagnostics? (like timeouts, deadlocks)
- Do you require explicit message ordering? (otherwise using the `IReliableConcurrentQueue` could help. This is not yet in the library, but it's a small change)

Could you please download, update and run the [Load Demo](https://github.com/loekd/ServiceFabric.PubSubActors/tree/master/ServiceFabric.PubSubActors.LoadDemo) and configure that [here](https://github.com/loekd/ServiceFabric.PubSubActors/blob/master/ServiceFabric.PubSubActors.LoadDemo/PublisherService/PackageRoot/Config/Settings.xml) and [here](https://github.com/loekd/ServiceFabric.PubSubActors/blob/master/ServiceFabric.PubSubActors.LoadDemo/SubscriberService/PackageRoot/Config/Settings.xml) to get a general idea about performance (in a controlled environment) on your side? 
(and let me know what comes out :) )

This tool also attempts to saturate the publishing of events, by using [multiple threads](https://github.com/loekd/ServiceFabric.PubSubActors/blob/master/ServiceFabric.PubSubActors.LoadDemo/PublisherService/PublisherService.cs#L106) to write.

I do know that there's a tipping point to reliable collections. After a certain threshold gets crossed you'll get nothing but timeout errors. These should show up in diagnostics.

**Comment by johnkattenhorn** [comment]
Thanks for the feedback @loekd - I'll review our code a let you know asap.

**Comment by johnkattenhorn** [comment]
@loekd - you have a view as to the ideal number of partitions replica's for the BrokerService ? I've notice we were using the pre-packaged one so I'm just trying out a custom version and reviewing our configuration settings.

Looking at `GetPartitionForMessageAsync` I'm wondering if I should balance the number of partitions by messageTypes and number of nodes.

**Comment by johnkattenhorn** [comment]
Hi,

We switched from the prepackaged service broker and created our own. I also the parameters `Period` and `DueTime` which we left at 200ms and 5 seconds respectively.

The chart below so a caparison run between the pre-packages and the custom broker, we the custom broker slightly slower for some reason.

<img width="913" alt="image" src="https://cloud.githubusercontent.com/assets/616249/22904794/a5f36064-f235-11e6-9638-b8e8eb07b844.png">

There no errors, timeouts reported through the callback, but almost immediately the response time of the publish jumps from milliseconds to seconds, see below a small snippet of our log:

```
13 Feb 2017 21:37:52.821
Queuing Packet from Device: [10.1.51.129] completed in 2356.8 ms 
13 Feb 2017 21:37:52.821
Publishing Packets for Device: [10.1.51.129] completed in 2049.6 ms
```
Publishing Packets sits inside Queuing Packet so you can see it eats a lot of total amount of time. It's a shame there's no way to fire and forget the publish.

I've also noticed to processing batches of exactly 100 from the queue like this

```
13 Feb 2017 21:37:58.910
BrokerService: [ProcessQueues - Processing 100 items from queue urn:M2M.Events.DataContracts.PacketSent_-115463902 for subscriber: fabric:/M2M.Kingfisher/EventHubService - ID:singleton - Guid:53946ca7-758a-45e4
```

I've also a fixed 3 second period in the code for the cancellationToken in `RunAsync` in BrokerService, could this account for the delay ?

```
 var cts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
```
The message is tiny, about 88 bytes in total on 1 message and there is a single message type being published from a stateful service to two stateless services.

We will continue to look into the library for a day or two and if we can't find some more performance we might have think again with our design. Any help or ideas would be greatly appreciated.

**Comment by johnkattenhorn** [comment]
I've branched the code and now I have this BrokerService declaration and I'm trying the load test again to see what effect it has:

```
            Period = TimeSpan.FromMilliseconds(200);
            DueTime = TimeSpan.FromSeconds(5);
            ProcessMessagePeriod = TimeSpan.FromSeconds(5);
            MaxDequeuesInOneIteration = 1000;
```

**Comment by loekd** [comment]
Using your own BrokerService is a good choice here. The prepackaged one is not configurable.


Adjusting `Period `increases the 'polling frequency' in the message processing loop. (sleep after processing)
The `DueTime `is the time to wait before starting the message processing loop.
The 3 second TimeSpan is the maximum duration one processing loop can take. (give or take)
(it should be a protected property, this is a bug)

In order to get a small amount of small messages across the wire in a short time, I'd try setting `Period ` really low, like 5 - 20 ms. 


The fact that you're always sending 100 message is an indication that processing is slower than ingestion. Adjust `MaxDequeuesInOneIteration` to influence the processing batch size.

Scaling out doesn't help here, using one message type means using one `BrokerService` replica.

If you want we can work together on implementing batch support for message delivery. That'll increase throughput for sure.

**Comment by johnkattenhorn** [comment]
@loekd - Thanks for the response. We've branched the code to expose the properties as you suggested.

We've increased the batch size already and we are now seeing that the batches are not very rarely hitting the batchsize limit but made little impact on the throughput.

Working on Batching together would be awesome and we are strong supporters of Open Source but we are but we are under a serious time pressure and we wouldn't be able to share the application code in the open as it's for a client project (although absolutely happy to share the results of the changes of the library we could come up with together).

You can PM me on john.kattenhorn@applicita.com if you would like to discuss this further.

**Comment by johnkattenhorn** [comment]
Quick update, we've temporary added the source to our solution and observed and made changes to the following:

We found that we were losing time by re-finding the servicebrokerurl over and over so we now cache this and supply it to the publish service override.

We found that `TimeoutRetryHelper` waited 200ms by default and we shorten this to 50 but we are not seeing any Timeouts.

We've instrument the `BrokerService` with `Serilog.Timings` to see where the time goes. 

We've learnt that the time publish includes the time for the subscribers to receive the events too, this makes sense since the state management has to be awaited.

This is the settings we are currently working with:

```
   public BrokerService(StatefulServiceContext context)
            : base(context)
        {
            var logger = ConfigureLogger();

            ServiceEventSourceMessageCallback = (message) => logger.Information("BrokerService: [{message}]", message);

            Period = TimeSpan.FromMilliseconds(1);
            DueTime = TimeSpan.FromSeconds(1);

            Logger = logger;

            MaxDequeuesInOneIteration = 1000;
            ProcessMessagePeriod = TimeSpan.FromMilliseconds(500);
        }
```

**Comment by johnkattenhorn** [comment]
We've implemented batching and are about to start load testing it

**Comment by johnkattenhorn** [comment]
So far in our load testing batching hasn't made much of a dent in the overall throughput, it's fast but just not fast enough for our needs, it appears that the ultimate performance comes from the ability to process queued messages.

BrokerService: Subscriber Publish for Queue: [M2M.Events.DataContracts.EventHubPacketSent_456681256]: DequeueBatchAsync [11] completed in 419.3 ms 

The publishing is very fast but having to iterate through the dequeue is pretty slow, shame that is no way to dequeue a block.

On further discussion we really need the publishing and subscribing of events to be "out-of-process" e.g. not tied together with StateManager updates, replication in a fire and forget type approach.

We are looking into using Pub / Sub to play a supporting role is registering event hubs and do the actual publishing of events through the eventhub.

**Comment by johnkattenhorn** [comment]
We've replaced the IReliableQueue implement with a IConcurrentQueue implementation (although it's still in preview) and the performance was increased 760%.

If you don't care about the order of the events (we don't) and don't mind the preview nature then this seems to prove that the basic code is fine, it's just the speed of the IReliableQueue.
